core/meterer/paymentlogic/paymentlogic.go

```diff
@@ -0,0 +1,155 @@
+//   - Charges apply to ALL allowed quorums, even if not all are used
+//   - On-demand requests have special requirements: they must use ETH and EIGEN quorums only
+func ValidateReservations(
+	reservations map[core.QuorumID]*core.ReservedPayment,
+	quorumConfigs map[core.QuorumID]*core.PaymentQuorumProtocolConfig,
+	quorumNumbers []uint8,
+	paymentHeaderTimestampNs int64,
+	receivedTimestampNs int64,
+) error {
+	reservationQuorums := make([]uint8, 0, len(reservations))
+	reservationWindows := make(map[core.QuorumID]uint64, len(reservations))
+	requestReservationPeriods := make(map[core.QuorumID]uint64, len(reservations))
+
+	// Gather quorums the user had an reservations on and relevant quorum configurations
+	for quorumID := range reservations {
+		reservationQuorums = append(reservationQuorums, uint8(quorumID))
+		_, ok := quorumConfigs[quorumID]
+		if !ok {
+			return fmt.Errorf("quorum config not found for quorum %d", quorumID)
+		}
+		reservationWindows[quorumID] = quorumConfigs[quorumID].ReservationRateLimitWindow
+		requestReservationPeriods[quorumID] = GetReservationPeriodByNanosecond(paymentHeaderTimestampNs, quorumConfigs[quorumID].ReservationRateLimitWindow)
+	}
+	if err := ValidateQuorum(quorumNumbers, reservationQuorums); err != nil {
+		return err
+	}
+	// Validate the used reservations are active and is of valid periods
+	for _, quorumID := range quorumNumbers {
+		reservation := reservations[core.QuorumID(quorumID)]
+		if !reservation.IsActiveByNanosecond(paymentHeaderTimestampNs) {
+			return errors.New("reservation not active")
+		}
+		if !ValidateReservationPeriod(reservation, requestReservationPeriods[quorumID], reservationWindows[quorumID], receivedTimestampNs) {
+			return fmt.Errorf("invalid reservation period for reservation on quorum %d", quorumID)
+		}
+	}
+
+	return nil
+}
+
+// ValidateReservationPeriod checks if the provided reservation period is valid
+// Note: This is called per-quorum since reservation is for a single quorum.
+func ValidateReservationPeriod(reservation *core.ReservedPayment, requestReservationPeriod uint64, reservationWindow uint64, receivedTimestampNs int64) bool {
+	currentReservationPeriod := GetReservationPeriodByNanosecond(receivedTimestampNs, reservationWindow)
+	// Valid reservation periods are either the current bin or the previous bin
+	isCurrentOrPreviousPeriod := requestReservationPeriod == currentReservationPeriod || requestReservationPeriod == (currentReservationPeriod-reservationWindow)
+	startPeriod := GetReservationPeriod(int64(reservation.StartTimestamp), reservationWindow)
+	endPeriod := GetReservationPeriod(int64(reservation.EndTimestamp), reservationWindow)
+	isWithinReservationWindow := startPeriod <= requestReservationPeriod && requestReservationPeriod < endPeriod
+	if !isCurrentOrPreviousPeriod || !isWithinReservationWindow {
+		return false
+	}
+	return true
+}
+
+func IsOnDemandPayment(paymentMetadata *core.PaymentMetadata) bool {
```

Comment:
Can we add a comment here for why this is considered on demand to document the current interpretation.